### PR TITLE
AdaptiveCardHost - Added "context" property

### PR DESCRIPTION
**Pull request by fabiofranzini**
| Q               | A
| --------------- | ---
| New feature?    | [X]
| Related issues?  | #1145 #1154

#### What's in this Pull Request?

Added SPFx "context" property to "AdaptiveCardHost" control to inject "@context" field in the data object for access context and theme information inside the Adaptive Card using the templating model.

### CODE REVIEW

File: src/controls/adaptiveCardHost/AdaptiveCardHostHelpers.ts [hunk 1]
```diff
@@ -1024,4 +1025,27 @@ const setCSSVariables = (domElement: HTMLElement, prefix: string, obj: any) => {
             domElement.style.setProperty(`--${prefix}-${key}`, obj[key]);
         });
     }
+};
+
+export const injectContextProperty = (dataObject: { "$root": object }, theme: ITheme, spfxContext: BaseComponentContext): IEvaluationContext => {
+    let evaluationContext: IEvaluationContext;
+    let context = {
+        theme: theme,
+        aadInfo: (spfxContext?.pageContext?.aadInfo) ? JSON.parse(JSON.stringify(spfxContext?.pageContext?.aadInfo)) : undefined,
```
Comment (AJIXuMuK) [inline src/controls/adaptiveCardHost/AdaptiveCardHostHelpers.ts:1034]: `JSON.stringify` on classes may lead to issues in future if there some new properties-instances with circular dependencies.
Not ideal way, but not buggy, is to directly specify what properties from the class you want to copy to your context object.

**Review by AJIXuMuK — CHANGES_REQUESTED** [review]

File: src/controls/adaptiveCardHost/AdaptiveCardHostHelpers.ts (hunk 1 above)
Comment (fabiofranzini) [reply src/controls/adaptiveCardHost/AdaptiveCardHostHelpers.ts:1034]: Hi @AJIXuMuK , I understand your request of changes and you are right. Let me understand a different way to do that, and I'll push the updates.

Comment (fabiofranzini) [reply src/controls/adaptiveCardHost/AdaptiveCardHostHelpers.ts]: Hi @AJIXuMuK , I followed your suggestion and fixed how to retrieve and use context properties. This way there are no problems with serialization. Thanks!

**Comment by fabiofranzini** [comment]
Hi @AJIXuMuK , I removed the "spPartialLegacyContext" from my previous commit because it's doesn't have sense at this moment. The legacy context from SPFx PageContext is legacy and at this moment it's not thought to use it in any scenario.

I also added a new commit with the code to add the "telemetry" for this control. 🙂

**Comment by fabiofranzini** [comment]
Hi @AJIXuMuK, can you check the last commit and give me some feedback on this code? I need to approve this PR before creating a new one 🙂

**Comment by fabiofranzini** [comment]
Hi everyone, I'm implementing Adaptive Cards designer as a react control. to do this I'm thinking about how to modify and reuse this control. This PR was to introduce "direct" support to SharePoint context properties, directly in the "data" object.
This, however, after much reasoning, introduces a "link" with the context of SPFx, and if one does not want to use it in this way one would be forced to have to pass the SharePoint context anyway.

That said, I was struggling to change the PR again and introduce not a control property, but a helper that allows you to inject the same properties in a way like this:

```
let dataObject = {"a":1};
<AdaptiveCardHost
  ...
  data={injectSPFxContextData(dataObject)}
  ...
/>
```
before approving this PR, can you tell me your opinion? 🙂
/cc @AJIXuMuK @joelfmrodrigues @joaojmendes

**Comment by AJIXuMuK** [comment]
@fabiofranzini - sorry for the long delay.
Maybe we could just make the `context` property optional?
You already have the logic to combine `data` and `context`.
If the `context` is optional, developers won't need to call any additional functions.

What do you think?

**Comment by fabiofranzini** [comment]
> @fabiofranzini - sorry for the long delay. Maybe we could just make the `context` property optional? You already have the logic to combine `data` and `context`. If the `context` is optional, developers won't need to call any additional functions.
> 
> What do you think?

Hi @AJIXuMuK, no problems 🙂
Yes, this could be a good compromise...
Let me do this change today and I'll notify you for the review 🙂

**Comment by fabiofranzini** [comment]
@AJIXuMuK done! 
I merged another fix from #1154 
Sorry for this, but I had already gone ahead with the change and then I found myself with the code already written.
For the next few times I will avoid doing this and will keep the code separate for each issue.
I have marked the "isUniqueControlInPage" property as obsolete and we no longer use it in the code, so those who have already used the control will not have problems compiling. 🙂

**Comment by AJIXuMuK** [comment]
@fabiofranzini - `context` is still marked as required in the docs.
Also - could you please explain the usage of `nano-css`?
Why is that needed instead of component's `scss` or something from Fluent UI?

**Comment by fabiofranzini** [comment]
@AJIXuMuK fixed the docs, thanks 🙂
I used `nano-css` to generate dynamically the class names for some styles to apply to the Adaptive Card HostConfig.
This is because the Adaptive Card HostConfig has a property called `cssClassNamePrefix`, and to set the rendering style (I need it mostly to solve the style problem in the Sub Actions menu) I need to have "dynamic" class names. Otherwise, if they were static, I would have the conflict problem if multiple instances of this control are on the same page (this is the way suggested by the Adaptive Cards product group to remove the `isUniqueControlInPage` property)

**Comment by AJIXuMuK** [comment]
merged!

**Comment by fabiofranzini** [comment]
> merged!

Thanks @AJIXuMuK!
Next step designer 😀

**Pull request merged**